Added 2 new endpoints for PlaylistsApi.

akshays2112 opened the pull request.

akshays2112 committed:
Merge pull request #1 from Ringobot/master

Sync

akshays2112 committed:
Small change to ShowsApi to add XML comment for AccessToken.
First 2 of the missing endpoints done for PlaylistApi.
Major changes here so will wait to resolve these before pushing on to quickly complete the rest of the missing endpoints.

akshays2112 committed:
No new line at end of file problem fixed.

akshays2112 committed:
Small fix for logic error in $"{}".

akshays2112 committed:
Did it another way and this is much much better.
No need to merge yet as I will complete the entire remaining endpoints before need to merge.

Request Daniel to verify if this method is good enough and then will proceed with the rest of the endpoints.

Since Json attributes and ShouldSerialize<PropName>() methods are used the need for the PutJsonString is gone so this code is removed.

Though personally I prefer PutJsonString for everything but I am old school. However for return values definitely support classes to be returned for end user to use of course of course.

akshays2112 commented:
Hi Daniel,

Requesting just a quick look at the way I solved the Dynamic Object/Class Type problem with Newtonsoft.Json ShouldSerialize<PropName>() method.

This is WIP.  Please approve of this and then I continue and finish of all endpoints first pass.  Then consolidate using your classes and return types (classes) from the other pull request of yours and that should save you time I think.  Well other than review time.

Thanks for your patience and time.

Best Regards,

Akshay Srinivasan

DanielLarsenNZ committed:
Refactor SpotifyWebApi POST Response

Change return type from HttpResponse to SpotifyResponse<T>. Deserialising body if any.

DanielLarsenNZ commented:
Hi Akshay,

You don't need `ShouldSerialize` at all. You can replace this object with my example here: https://github.com/Ringobot/SpotifyApi.NetCore/pull/40/files#diff-0980c5ae9847c84bd559259bfae5ab7eR7

Notice the use of `NullValueHandling` attribute arguments here.

I have also changed `SpotifyWebApi` to return a new `SpotifyResponse` instead of `HttpResponseMessage` from `Post` methods. I have added a new generic overload so that you can return a deserialized object from `Post<T>()`. I have updated your PlaylistApi implementation to use this new helper.

DanielLarsenNZ commented:
For the release notes:

This release changes the return types of `Post()` and `Put()` from `HttpResponseMessage` to `SpotifyResponse<T>`. No internal callers were using the return type so this is not a breaking change. A new generic `Post<T>()` internal helper has been added to deserialize POST response bodies to an object typed `T`

`Description` is added to `PlaylistSimplified`

DanielLarsenNZ commented on src/SpotifyApi.NetCore/Models/Playlist.cs:
```diff
@@ -269,4 +269,16 @@ public partial class PlaylistPaged
+    /// <summary>
+    /// Upon adding items to a playlist a snapshot_id is returned.
+    /// </summary>
+    public partial class PlaylistSnapshotID
```
Please create a separate class file for this and rename to `ModifyPlaylistResponse`. See this example: https://github.com/Ringobot/SpotifyApi.NetCore/pull/40/files#diff-3070e73922144bda6f13360673a6810cR5

DanielLarsenNZ commented on src/SpotifyApi.NetCore/IPlaylistsApi.cs:
```diff
@@ -118,5 +118,67 @@ public interface IPlaylistsApi
+        /// <remarks>
+        /// https://developer.spotify.com/documentation/web-api/reference/playlists/add-tracks-to-playlist/
+        /// </remarks>
+        Task<T> AddItemsToPlaylist<T>(
+            string playlistId,
+            string[] uris,
+            int? position = null,
+            string accessToken = null
+            );
```
We don't need a generic method for this one, thanks.

DanielLarsenNZ commented on src/SpotifyApi.NetCore/IPlaylistsApi.cs:
```diff
@@ -118,5 +118,67 @@ public interface IPlaylistsApi
+        /// </remarks>
+        Task ChangePlaylistDetails(
+            string playlistId,
+            string name = null,
+            bool? makePublic = null,
+            bool? collaborative = null,
+            string description = null,
```
Please use a request DTO here, see https://github.com/Ringobot/SpotifyApi.NetCore/pull/40/files#diff-bccd16b25f81b67901713ae0c33510e0R152

DanielLarsenNZ commented on src/SpotifyApi.NetCore/PlaylistsApi.cs:
```diff
@@ -194,5 +196,132 @@ public async Task<T> GetPlaylist<T>(string playlistId, string accessToken = null
+        /// <remarks>
+        /// https://developer.spotify.com/documentation/web-api/reference/playlists/add-tracks-to-playlist/
+        /// </remarks>
+        public async Task<T> AddItemsToPlaylist<T>(
```
Generic method is not needed here. Should return `ModifyPlaylistResponse`

DanielLarsenNZ commented on src/SpotifyApi.NetCore/PlaylistsApi.cs:
```diff
@@ -194,5 +196,132 @@ public async Task<T> GetPlaylist<T>(string playlistId, string accessToken = null
+        #endregion
+
+        #region ChangePlaylistDetails
+        class PutDataPayloadPlaylistDetails
```
Please don't nest classed. Create a new class. See https://github.com/Ringobot/SpotifyApi.NetCore/pull/40/files#diff-0980c5ae9847c84bd559259bfae5ab7eR5

DanielLarsenNZ commented on src/SpotifyApi.NetCore/PlaylistsApi.cs:
```diff
@@ -194,5 +196,132 @@ public async Task<T> GetPlaylist<T>(string playlistId, string accessToken = null
+                Description = description;
+            }
+
+            public bool ShouldSerializeName()
+            {
+                return !string.IsNullOrEmpty(Name);
+            }
+
+            public bool ShouldSerializePublic()
+            {
+                return Public != null;
+            }
+
+            public bool ShouldSerializeCollaborative ()
+            {
+                return Collaborative != null;
+            }
+
+            public bool ShouldSerializeDescription()
+            {
+                return !string.IsNullOrEmpty(Description);
+            }
```
Don't do this. See https://github.com/Ringobot/SpotifyApi.NetCore/pull/40/files#diff-0980c5ae9847c84bd559259bfae5ab7eR7

DanielLarsenNZ committed:
Refactor SpotifyWebApi POST Response

Change return type from HttpResponse to SpotifyResponse<T>. Deserialising body if any.

akshays2112 commented:
Hi Daniel,

Super. ♥ all of it. Will do it. Drinking right now (11:30PM here in Inja) so will do it tomorrow when sober.

Thanks for your patience and time.

Best Regards,

Akshay Srinivasan

akshays2112 committed:
Daniel, I need your help with CreatePlaylist.  Please check out why the return is not deserializing to Playlist model.  I cannot find the mistake.  Also the CreatePlaylist is creating the test playlist just fine as you can see.  It is just a deserialization problem of the returned Playlist model of the new playlist that was successfully created.

akshays2112 committed:
Merge branch 'master' of https://github.com/akshays2112/SpotifyApi.NetCore

akshays2112 committed:
Daniel please help.
When I synced I got a new copy of the new post which returns SpotifyResponse<T> so I corrected my code to work with this.  Please check this code too.  My understanding of generics is iffy.

DanielLarsenNZ committed:
CreatePlaylist use PlaylistDetails

DanielLarsenNZ committed:
ChangePlaylistDetails now uses PlaylistDetails

DanielLarsenNZ committed:
ChangePlaylistDetails now uses PlaylistDetails

DanielLarsenNZ committed:
ChangePlaylistDetails now uses PlaylistDetails

DanielLarsenNZ reviewed (commented):
How's this Akshay?

akshays2112 commented:
Hi Daniel,

You are an awesome coder my friend.  Yes, basically that damn deserializing bug was due to it being Paged<T> and not a regular thing.  Rest of all the changes are awesome.  Yes, this time I read through each and every one.

I think we can merge it if you are ok with it.
Then you can merge in your other pull request as you deem fit as usual.
Then I can sync up my stuff and continue closing out the remaining few endpoints easily.

Thanks for your patience and time.

Best Regards,

Akshay Srinivasan

akshays2112 commented:
Hi Daniel,

 

Went through all your changes. 100% approved.  Please merge and close out all pull requests and then I will sync and continue soon as I get word from you.  Ya the life issues were well life problems not health problems LOL.  Meh!! Sh*t happens you know LOL!

 

Cheers,

 

Akshay Srinivasan

+91-9884967112

Mumbai, India

 

From: Daniel Larsen <notifications@github.com> 
Sent: Monday, August 3, 2020 3:03 PM
To: Ringobot/SpotifyApi.NetCore <SpotifyApi.NetCore@noreply.github.com>
Cc: Akshay Srinivasan <akshays2112@gmail.com>; Author <author@noreply.github.com>
Subject: Re: [Ringobot/SpotifyApi.NetCore] Added 2 new endpoints for PlaylistsApi. (#41)

 

@DanielLarsenNZ commented on this pull request.

How's this Akshay?

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub <https://github.com/Ringobot/SpotifyApi.NetCore/pull/41#pullrequestreview-459884648> , or unsubscribe <https://github.com/notifications/unsubscribe-auth/AIRJIGVHWJTYEV44AIF5MJ3R6Z76TANCNFSM4PBJ3QRQ> .  <https://github.com/notifications/beacon/AIRJIGSZP5YKJT7WNRYHUDLR6Z76TA5CNFSM4PBJ3QR2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGODNUUQ2A.gif>
